File: Datastructures/stack/stack.py
'''
Implementation of stack using list.

Operations are O(1)
'''

class Stack():

    def __init__(self):
        self.items = []

    def is_empty(self):
        return self._is_empty()

    def push(self, item):
        self._push(item)

    def pop(self):
        return self._pop()

    def peek(self):
        return self._peek()

    def size(self):
        return self._size()

    def print_stack(self):
        print self.items

    # privates

    def _is_empty(self):
        return self.items == []

    def _push(self, item):
        self.items.append(item)

    def _pop(self):
        if not self.is_empty():
            return self.items.pop()
        raise Exception('Stack is already empty')

    def _peek(self):
        return self.items[len(self.items)-1]

    def _size(self):
        return len(self.items)

if __name__ == "__main__":
    stack = Stack()
    stack.push('kuldeep')
    stack.print_stack()
    stack.push(3)
    stack.print_stack()
    stack.pop()
    stack.print_stack()
    stack.push('test')
    stack.peek()
    stack.print_stack()

File: Datastructures/stack/stack.go
/*
Implementation of stack in go language using slices.
*/

package main

import (
        "fmt"
        "errors"
       )


type stack struct{
    list []int
}


func (s stack) is_empty() bool{
    return len(s.list) == 0
}


func (s stack) peek() (int, error){
    if !s.is_empty(){
        return s.list[len(s.list)-1], nil
    }
    return s.size(), errors.New("stack is empty")
}


func (s stack)size() int{
    return len(s.list)
}


func (s *stack)push(value int){
    s.list = append(s.list, value)
}


func (s *stack)push(value string){
    s.list = append(s.list, value)
}


func (s *stack)pop() []int{
    return s.list[:len(s.list)-1]
}


func (s stack)print_stack() {
    fmt.Println(s.list)
}


func main() {
    s := stack{}

    v := fmt.Println 
    v(s)
    v(s.size())
    v(s.is_empty())
    v(s.peek())
    s.push(3)
    s.push(4)
    s.push(5)
    s.print_stack()
}
